### PR TITLE
refactor(minor-ampd): separate generic public key for multisig signing and cosmos public key for tx signing

### PR DESCRIPTION
**Pull request by fish-sammy**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: ampd/src/tofnd/grpc.rs [hunk 1]
```diff
@@ -90,7 +87,7 @@ impl Multisig for MultisigClient {
         &self,
         key_uid: &str,
         data: MessageDigest,
-        pub_key: &PublicKey,
+        pub_key: PublicKey,
```
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/tofnd/grpc.rs:90]: https://github.com/axelarnetwork/axelar-amplifier/blob/f7b3c08786dd5e13c47a2a0b5f68bb23270a95af/ampd/src/broadcaster/mod.rs#L242 cuz here a future cannot take a reference of a temporary variable.

File: ampd/src/types/key.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    #[error("invalid raw bytes")]
+    InvalidRawBytes,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum PublicKey {
+    Secp256k1(k256::ecdsa::VerifyingKey),
+    Ed25519(ed25519_dalek::VerifyingKey),
+    // TODO: add support for Schnorr
+    Schnorr,
```
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/types/key.rs:25]: sure

File: ampd/src/types/key.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    }
+}
+
+impl TryFrom<multisig::key::PublicKey> for PublicKey {
+    type Error = Report<Error>;
+
+    fn try_from(key: multisig::key::PublicKey) -> Result<Self> {
+        (&key).try_into()
+    }
+}
```
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/types/key.rs:90]: Not really, but good to have for convenience.

File: ampd/src/types/key.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+impl TryFrom<PublicKey> for CosmosPublicKey {
+    type Error = Report<Error>;
+
+    fn try_from(key: PublicKey) -> Result<Self> {
+        (&key).try_into()
+    }
+}
```
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/types/key.rs:114]: same here.

File: ampd/src/types/key.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        .expect("must be valid cosmos key")
+    }
+}
+
+impl From<CosmosPublicKey> for PublicKey {
+    fn from(key: CosmosPublicKey) -> Self {
+        (&key).into()
+    }
+}
```
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/types/key.rs:131]: same here

File: ampd/src/types/key.rs
[the remark this answers is not in the document]
Comment (fish-sammy) [reply ampd/src/types/key.rs:1]: Ideally, `cosmrs::tendermint::PublicKey` and `cosmrs::crypto::PublicKey` should never be used after this PR in favour of types::key::PublicKey. I might be able to get rid of `multisig::key::PublicKey` though, in a separate PR. `KeygenResponse::PubKey` is generated from the protobuf so I am not sure what I can do about it.

**Review by cgorenflo — APPROVED** [review]

**Pull request merged**